### DISCUSSION TITLE
Generative AI megathread

### DISCUSSION THREAD

**Discussion by teolemon** [openfoodfacts/openfoodfacts-ai#272] [Show and tell]
### What
- Investigate writing a script to augment taxonomy files using the ChatGPT API
- Prompts like: 
`translate this in as many languages as possible: en:PE - Polyethylene, Polyethylene`
`can you format the list with the format "2 letter langcode:translation"`
"Summarise in 3 sentences the production of material X, how it is used in food packaging and and how it is recycled"
might be useful
- This would be particularly useful for Open Products Facts 26000 category system (from the UNSPC) which is currently in English, without synonyms, without links to Wikidata, and without translations
- Given the sheer amount of content to review, we'd need a system to do that entry per entry, create a git branch, commit, and create a PR.
- Alex has proposed a Taxonomy Editor integration, but that's longer term

### Where
- Taxonomy Editor (https://github.com/openfoodfacts/taxonomy-editor/issues/233)
  - Schema definition: https://openfoodfacts.slack.com/archives/C03MEM1CMJ4/p1676631826269169
  - Subcategories: https://openfoodfacts.slack.com/files/U02KZKKA5/F04Q4ENDBR8/image.png
- IDE used for taxonomy edition
- Odoo Connect (https://twitter.com/PaulUheet/status/1630203943140687874?t=qbwmOWfIu9ZNIWYBk0bgYw)

### How
- Short term: Using the closed/paid for ChatGPT Python API
- Migrating to a fully open stack ASAP

### Example for adopting our taxonomy format
```
I would like you to output all your answers using the following format:<2 letter language code:name of the parent in English
2 letter language code:name of the item in said language code, synonym one in the same language, synonym two in the same language, …
description:2 letter language code:a three line description of the item in said language code
wikidata:en:Qid of the item on Wikidata if you know it, otherwise nothing

create an entry for chicken tikka in the maximum of languages
```

### Generating a data model for Open Products Facts

`
Can you list as bullet points the relevant data fields to describe "Molecular diagnostics analyzer" for a product datasheet.
`

- Product Name: Molecular Diagnostics Analyzer
- Application: [specify the type of sample analysis, such as clinical or research]
- Technology: [indicate the molecular diagnostic technology used, such as PCR, NGS, etc.]
- Sample type compatibility: [list the types of samples that can be analyzed, such as blood, saliva, etc.]
- Detection methods: [indicate the types of targets the analyzer can detect, such as DNA, RNA, pathogens, etc.]
- Throughput: [indicate the number of samples that can be processed per unit of time, such as per hour or day]
- Run time: [specify the total time required for a sample analysis, from sample preparation to result generation]
- Sensitivity: [indicate the minimum concentration of the target that can be detected]
- Specificity: [indicate the degree of accuracy in detecting the target of interest, as well as avoiding false positive results]
- Data management: [indicate the type of data storage and analysis capabilities, such as cloud storage, real-time data analysis, etc.]
- User interface: [describe the type of interface for sample analysis, such as touchscreen or computer-based]
- Power source: [indicate the type of power required, such as AC power or battery-operated]
- Dimensions: [provide the physical dimensions of the analyzer, such as height, width, and depth]
- Weight: [indicate the weight of the analyzer for ease of transport and handling]
- Brand: [name of the brand producing the product]
- SKU or product code: [unique identifier for the product]
- MSRP: [manufacturer's suggested retail price].

Can you list as bullet points the relevant data fields to describe "Microphone" for a product datasheet.

Sure! Here are some relevant data fields to describe a microphone for a product datasheet:

Type of microphone (e.g. dynamic, condenser, ribbon, USB)
Polar pattern (e.g. cardioid, omnidirectional, bidirectional)
Frequency response (range and sensitivity)
Sensitivity (output level for a given sound pressure level)
Impedance (resistance to electrical current flow)
Signal-to-noise ratio (ratio of the desired signal to background noise)
Maximum SPL (maximum sound pressure level before distortion occurs)
Connector type (e.g. XLR, USB, TRS)
Power supply requirements (e.g. phantom power, battery, USB)
Size and weight of the microphone
Accessories included (e.g. shock mount, windscreen, carrying case)
Note that some of these fields may not apply to all types of microphones, and additional fields may be relevant depending on the specific microphone and intended use.

### Generating subcategories for Open Products Facts
`Provide a bullet point list of subcategories for "Inflatable toys"`

- Inflatable pools
- Inflatable water slides
- Inflatable bouncers and castles
- Inflatable rafts and boats
- Inflatable beach balls
- Inflatable air mattresses
- Inflatable loungers and chairs
- Inflatable snow tubes
- Inflatable kayaks
- Inflatable pool toys (dartboards, basketball hoops, etc.)

### Finding vegan status from ingredients

is this ingredient list vegan and why ?  Lait entier (60,8%), sucre, crème (lait, beurre de cacao, pâte de cacao, poudre de lait écrémé, chocolat noir (2,7%) [pâte de cacao, sucre, émulsifiant (lécithine de tournesol), arốme naturel de vanille], cacao maigre en poudre, caramel (sucre, eau, sirop de glucose, correcteur d'acidité (hydroxyde de sodium)], cacao en poudre, lactose (lait), gélatine bovine, émulsifiant (E471), sel, arôme
No, this ingredient list is not vegan. The product contains milk (lait entier), cream (crème), milk powder (poudre de lait écrémé), and gelatine derived from bovine (gélatine bovine), which are animal-based products and not suitable for a vegan diet.

### Finding gluten free status from ingredients
is this ingredient list gluten free and why ?  Lait entier (60,8%), sucre, crème (lait, beurre de cacao, pâte de cacao, poudre de lait écrémé, chocolat noir (2,7%) [pâte de cacao, sucre, émulsifiant (lécithine de tournesol), arốme naturel de vanille], cacao maigre en poudre, caramel (sucre, eau, sirop de glucose, correcteur d'acidité (hydroxyde de sodium)], cacao en poudre, lactose (lait), gélatine bovine, émulsifiant (E471), sel, arôme

It is not specified in the ingredient list whether the product is gluten-free. The listed ingredients do not contain gluten, but it is always advisable to check with the manufacturer or check for a certified gluten-free label to ensure that the product has not been cross-contaminated with gluten during processing or manufacturing.

### Finding halal status from ingredients
is this ingredient list halal and why ?  Lait entier (60,8%), sucre, crème (lait, beurre de cacao, pâte de cacao, poudre de lait écrémé, chocolat noir (2,7%) [pâte de cacao, sucre, émulsifiant (lécithine de tournesol), arốme naturel de vanille], cacao maigre en poudre, caramel (sucre, eau, sirop de glucose, correcteur d'acidité (hydroxyde de sodium)], cacao en poudre, lactose (lait), gélatine bovine, émulsifiant (E471), sel, arôme

This ingredient list is not necessarily Halal certified. The presence of bovine gelatine (gélatine bovine) raises concerns for those who follow a Halal diet as it may not be sourced from a Halal-certified animal. It is advisable to check with the manufacturer or look for a Halal certification label to determine whether the product is suitable for a Halal diet.

### Finding kosher status from ingredients
is this ingredient list kosher and why ?  Lait entier (60,8%), sucre, crème (lait, beurre de cacao, pâte de cacao, poudre de lait écrémé, chocolat noir (2,7%) [pâte de cacao, sucre, émulsifiant (lécithine de tournesol), arốme naturel de vanille], cacao maigre en poudre, caramel (sucre, eau, sirop de glucose, correcteur d'acidité (hydroxyde de sodium)], cacao en poudre, lactose (lait), gélatine bovine, émulsifiant (E471), sel, arôme

This ingredient list is not necessarily Kosher certified. The presence of bovine gelatine (gélatine bovine) raises concerns for those who follow a Kosher diet as it may not be sourced from a Kosher-certified animal. It is advisable to check with the manufacturer or look for a Kosher certification label to determine whether the product is suitable for a Kosher diet.

###  lowercase ingredient list 
Can you lowercase the following ingredient list ? *JAMBON DE PORC (84% - ORIGINE FRANCE), *COUENNE DE PORC (9% - ORIGINE FRANCE), EAU, SEL DE MER, *SUCRE DE CANNE, #AROME NATUREL DE CITRON, #ARÔME NATUREL DE CAROUBE (PRÉSENCE NATURELLE DE POLYPHENOLS), *EXTRAIT DE LEVURE, ANTIOXYDANT: E300, #CAROTTE DÉSHYDRATÉ, *OIGNON DÉSHYDRATÉ, *POIREAU DÉSHYDRATÉ, #AIL DÉSHYDRATE, *ROMARIN DÉSHYDRATÉ, *POIVRE NOIR MOULU, AUXILIAIRE TECHNOLOGIQUE: ANTIAGGLOMÉRANT: E500,
*jambon de porc (84% - origine france), *couenne de porc (9% - origine france), eau, sel de mer, *sucre de canne, #arome naturel de citron, #arôme naturel de caroube (présence naturelle de polyphenols), *extrait de levure, antioxidant: e300, #carotte déshydraté, *oignon déshydraté, *poireau déshydraté, #ail déshydraté, *romarin déshydraté, *poivre noir moulu, auxiliaire technologique: antiagglomérant: e500.

### Correcting ingredient list
Can you correct the following ingredient list ? *JAMBON DE PORC (84% - ORIGINE FRANCE), *COUENNE DE PORC (9% - ORIGINE FRANCE), EAU, SEL DE MER, *SUCRE DE CANNE, #AROME NATUREL DE CITRON, #ARÔME NATUREL DE CAROUBE (PRÉSENCE NATURELLE DE POLYPHENOLS), *EXTRAIT DE LEVURE, ANTIOXYDANT: E300, #CAROTTE DÉSHYDRATÉ, *OIGNON DÉSHYDRATÉ, *POIREAU DÉSHYDRATÉ, #AIL DÉSHYDRATE, *ROMARIN DÉSHYDRATÉ, *POIVRE NOIR MOULU, AUXILIAIRE TECHNOLOGIQUE: ANTIAGGLOMÉRANT: E500,
Here is the corrected ingredient list:

*Jambon de porc (84%, origine France)
*Couenne de porc (9%, origine France)
Eau
Sel de mer
*Sucre de canne
#Arome naturel de citron
#Arôme naturel de caroube (présence naturelle de polyphénols)
*Extrait de levure
Antioxydant: E300
#Carotte déshydratée
*Oignon déshydraté
*Poireau déshydraté
#Ail déshydraté
*Romarin déshydraté
*Poivre noir moulu
Auxiliaire technologique: Antiagglomérant: E500

### Test with additives
is E471 vegan ?
The emulsifier E471 is typically made from either glycerol esters of fatty acids from plant origin or from glycerol esters of fatty acids from animal origin, such as pork fat. Whether or not E471 is considered vegan depends on the specific source of the fatty acids used to make it. Some vegans may consider E471 made from plant-based sources to be acceptable, while others may not consider it vegan regardless of its origin. It is best to check with the manufacturer of the product to determine the source of E471.

is E120 safe ?
E120, also known as cochineal or carmine, is a natural red food coloring derived from insects. It is considered safe for consumption by most people, but some individuals may have an allergic reaction to it. Additionally, the use of E120 is not considered suitable for vegans or vegetarians, as it is derived from insects. The European Food Safety Authority (EFSA) has established an acceptable daily intake (ADI) for E120, and considers it safe for use in food within the recommended limits.

#### Snap a photo and get step by step recipe suggestions
- https://twitter.com/sudu_cb/status/1636080774834257920

#### Just point your camera at the ingredients and generate a recipe!
https://twitter.com/JustMeDevin/status/1620724801190105088

#### Getting all supermarkets/restaurants for a country
- https://twitter.com/mirbole01/status/1614583332360036354

### Generating tweets based on specific categories / tags…
https://twitter.com/Lukaesch/status/1482087257553133570
Eg: Did you know that …

### Databases of prompts for ChatGPT/LLMs and Dall-E/Midjourney
https://promptbase.com/prompt/food-recipe-generator
https://promptbase.com/prompt/cartoon-vegetables-chibi-style

### Recognizing photos and generating description of the dish
- https://wikifoodia.vercel.app/#about

### Association mets vins
https://twitter.com/laerte_olmelli/status/1632517096700624901

### App recognizing nutrition label and ingredients
-https://twitter.com/Xbond49/status/1385695418378604545

**Comment by teolemon** [comment]
`
Can you list as bullet points the relevant data fields to describe "Molecular diagnostics analyzer" for a product datasheet.
`


- Product Name: Molecular Diagnostics Analyzer
- Application: [specify the type of sample analysis, such as clinical or research]
- Technology: [indicate the molecular diagnostic technology used, such as PCR, NGS, etc.]
- Sample type compatibility: [list the types of samples that can be analyzed, such as blood, saliva, etc.]
- Detection methods: [indicate the types of targets the analyzer can detect, such as DNA, RNA, pathogens, etc.]
- Throughput: [indicate the number of samples that can be processed per unit of time, such as per hour or day]
- Run time: [specify the total time required for a sample analysis, from sample preparation to result generation]
- Sensitivity: [indicate the minimum concentration of the target that can be detected]
- Specificity: [indicate the degree of accuracy in detecting the target of interest, as well as avoiding false positive results]
- Data management: [indicate the type of data storage and analysis capabilities, such as cloud storage, real-time data analysis, etc.]
- User interface: [describe the type of interface for sample analysis, such as touchscreen or computer-based]
- Power source: [indicate the type of power required, such as AC power or battery-operated]
- Dimensions: [provide the physical dimensions of the analyzer, such as height, width, and depth]
- Weight: [indicate the weight of the analyzer for ease of transport and handling]
- Brand: [name of the brand producing the product]
- SKU or product code: [unique identifier for the product]
- MSRP: [manufacturer's suggested retail price].

**Comment by teolemon** [comment]
`Provide a bullet point list of subcategories for "Inflatable toys"`

- Inflatable pools
- Inflatable water slides
- Inflatable bouncers and castles
- Inflatable rafts and boats
- Inflatable beach balls
- Inflatable air mattresses
- Inflatable loungers and chairs
- Inflatable snow tubes
- Inflatable kayaks
- Inflatable pool toys (dartboards, basketball hoops, etc.)

### Finding category from nutrition facts

### Finding category from ingredients

### Finding vegan status from ingredients

**Comment by teolemon** [comment]
![image](https://user-images.githubusercontent.com/1689815/217609288-17db0b7b-2611-44c1-b045-2fc0e4756911.png)

**Comment by teolemon** [comment]
![image](https://user-images.githubusercontent.com/1689815/217609336-8f745a45-b460-4688-8393-c2e0023a2320.png)

**Reply by teolemon** [reply]
https://github.com/openfoodfacts/openfoodfacts-server/pull/8084/files

**Comment by teolemon** [comment]
![image](https://user-images.githubusercontent.com/1689815/217609959-8a596137-82fc-490d-8117-d6e8ce094eab.png)

**Comment by teolemon** [comment]
![image](https://user-images.githubusercontent.com/1689815/217626561-c11e8ca9-6201-4008-b56c-c91a546127ec.png)

**Comment by teolemon** [comment]
![image](https://user-images.githubusercontent.com/1689815/221598908-54b38db7-a0e0-48d1-899f-2228efe44a70.png)
![image](https://user-images.githubusercontent.com/1689815/221598944-c8abdd7c-3fac-454c-8df5-34198cdc22b0.png)

**Comment by teolemon** [comment]
![image](https://user-images.githubusercontent.com/1689815/221599026-2b4670f0-993b-4590-9a60-c7df13dfb63d.png)
![image](https://user-images.githubusercontent.com/1689815/221599078-aa846014-8479-452c-913d-bcd67b722c1b.png)

**Reply by teolemon** [reply]
https://world.openfoodfacts.org/product/0080948630439/north-star-trading-company-snacks-nuts-heart-healthy

**Comment by benbenben2** [comment]
# Put all categories in plural 
![image](https://user-images.githubusercontent.com/110821832/225748737-22bf04cb-b944-4b88-9e50-efe71077927c.png)
![image](https://user-images.githubusercontent.com/110821832/225748767-557ad8d2-4112-4363-8b96-97fa761149a9.png)

**Comment by benbenben2** [comment]
Using ChatGPT to complete entries in different languages from wikidata.
**warning:** ChatGPT seems to have discrepancies when it comes to wikidata code and their content
![image](https://user-images.githubusercontent.com/110821832/226125741-4dbc6271-d84d-4614-84f7-79a8b5ff0095.png)
